keps/sig-scheduling/3101-pod-to-node-assignment/README.md

```diff
@@ -0,0 +1,944 @@
+    tests on pod admission. This will prevent users from working around the
+    scheduler to avoid coexist policy enforcement.
+-   If no Node was found that satisfies the filter for the incoming pod,
+    [Pod Preemption](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/#preemption)
+    may evict lower priority pods to satisfy the coexist policy of the incoming
+    pod.
+
+### API
+
+A new API field `pod.spec.coexistPolicy` is introduced to be used by the new
+filter. The value for the newly introduced field is an enum that has one of the
+following values: `DaemonsetAndStaticPods`, `Any` (Default) where `Any` is the
+same as the current behaviour and `DaemonsetAndStaticPods` would allow placing
+only a single pod on a node excluding daemonset and static pods, effectively
+enforcing 1:1 workload pod to node mapping.
+
+```yaml
+...
+spec:
+  coexistPolicy: <string>
+```
+
+This API will look as follows in Golang
+
+```go
+type PodSpec struct {
+  // CoexistPolicy specifies which other pods this pod can "coexist" with. It is similar to pod anti-affinity but
+  // limited to node topology and doesn't operate on pod labels, and so more performant but has narrow use
+  // cases defined by the hardcoded policies.
+  CoexistPolicy CoexistPolicy
+}
+
+// +enum
+type CoexistPolicy string
+
+const (
+    // Pod tolerates co-existing with other pods.
+    CoexistPolicyAny CoexistPolicy = "Any"
+    // Pod tolerates coexisting with daemonset and static pods only.
+    CoexistPolicyDaemonsetAndStaticPods CoexistPolicy = "DaemonsetAndStaticPods"
+)
+```
+
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy. Anything
+that would count as tricky in the implementation, and anything particularly
+challenging to test, should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+Unit and integration tests verifying the following
+
+-   a single pod is assigned to a node while allowing daemonsets and static
+    pods.
+-   scheduler doesn't schedule new pods to a node that already have a pod with
+    `DaemonsetAndStaticPods` policy.
+-   Daemonset and static pods can be scheduled to a node that has a pod with
+    `DaemonsetAndStaticPods` policy.
+-   kubelet rejects a pod if it was assigned directly to a node already running
+    a pod with `DaemonsetAndStaticPods` policy
+-   the API is no-op if feature gate is turned off
+
+Benchmark Tests
```

Comment:
this will not be worse than inter-pod anti-affinity, but I feel that there something can be validated on how the update will work for such pods. If pod is being updated as part of the ReplicaSet, and it takes some time to be torn down, would we ever want to allow scheduling on the same Node while the pod is being shutdown? Also it would be nice to prevent races when some other pod is scheduled on the Node blocking the updated pod with `DaemonsetAndStaticPods` be scheduled on that same node.

Reply:
I don't think we want to bake any logic specific to this feature regarding the two issues below. 

> would we ever want to allow scheduling on the same Node while the pod is being shutdown?

As long as the pod is not yet deleted, it will be taken into account. Similar to how we deal with normal resources like cpu and memory for pods shutting down.

> Also it would be nice to prevent races when some other pod is scheduled on the Node blocking the updated pod with DaemonsetAndStaticPods be scheduled on that same node.

If the pod with DaemonsetAndStaticPods has higher priority, it will kick out the other pods that raced into the node while the replicaset is being updated, otherwise the node is fair game for other pods.

Reply:
yes, I understand how it works. But since this repellent condition is super strong I wonder if any adjustments might be made to simplify rollout updates. Kicking off other pods will be quite long operation. Thinking more about it, I guess in many scenarios nodes will be specially marked to handle these DaemonsetAndStaticPods pods. And other pods may be repelled from them.